base/cvd/cuttlefish/common/libs/utils/files.cpp

```diff
@@ -81,6 +81,60 @@ bool FileExists(const std::string& path, bool follow_symlinks) {
   return (follow_symlinks ? stat : lstat)(path.c_str(), &st) == 0;
 }
 
+Result<dev_t> FileDeviceId(const std::string& path) {
+  struct stat out;
+  CF_EXPECTF(
+      stat(path.c_str(), &out) == 0,
+      "stat() failed trying to retrieve device ID information for \"{}\" "
+      "with error: {}",
+      path, strerror(errno));
+  return out.st_dev;
+}
+
+Result<bool> CanHardLink(const std::string& path1, const std::string& path2) {
+  return CF_EXPECT(FileDeviceId(path1)) == CF_EXPECT(FileDeviceId(path2));
+}
+
+Result<ino_t> FileInodeNumber(const std::string& path) {
+  struct stat out;
+  CF_EXPECTF(
+      stat(path.c_str(), &out) == 0,
+      "stat() failed trying to retrieve inode num information for \"{}\" "
+      "with error: {}",
+      path, strerror(errno));
+  return out.st_ino;
+}
+
+Result<bool> AreHardLinked(const std::string& path1, const std::string& path2) {
+  return CF_EXPECT(FileInodeNumber(path1)) == CF_EXPECT(FileInodeNumber(path2));
```

Comment:
inode numbers are only guaranteed to be unique within a single filesystem, so it's the same file if both `st_dev` and `st_ino` match.

(there might also be other conditions for it being the same file like bind mounting, not sure)

Reply:
You are right, we need the combination of both.  Expanded the conditional